sound/soc/sof/topology.c

```diff
@@ -2199,6 +2193,22 @@ static int sof_complete(struct snd_soc_component *scomp)
 			}
 	}
 
+	/*
+	 * finally update all widget IPC structures. If any of the ipc_setup callbacks fail, the
+	 * topology will be removed and all widgets will be unloaded resulting in freeing all
+	 * associated memories.
+	 */
+	list_for_each_entry(swidget, &sdev->widget_list, list) {
+		if (swidget->id != snd_soc_dapm_scheduler && widget_ops[swidget->id].ipc_setup) {
+			ret = widget_ops[swidget->id].ipc_setup(swidget);
```

Comment:
can this be done right after after `sof_set_widget_pipeline()` call above or this needs to be done for widgets not part of any pipeline? DO we even have such a widget?

just too many `list_for_each_entry(swidget, &sdev->widget_list, list)` in sequence.

It might result too deep indentation?